File: webapp/src/Websocket/types/EnvironmentSensors.ts
export type EnvironmentSensorRequestType =
    "getAllEnvironmentSensorData"
    | "uploadFirmware"
    | "scheduleTimeAdjustment"
    | "cancelTimeAdjustment"
    | "scheduleFirmwareUpgrade"
    | "cancelFirmwareUpgrade"
    | "adjustSleepTimeInSeconds"
    | "scheduleReset"
    | "cancelReset";

export type EnvironmentSensorRequest = {
    type: EnvironmentSensorRequestType;
    sensorId?: number;
    firmwareFilename?: string;
    firmwareBased64Encoded?: string;
    sleepTimeInSecondsDelta?: number;
}

export type FirmwareUpgradeState = {
    firmwareSize: number;
    offsetRequested: number;
    timestampDelta: number;
    receivedAt: number;
    rssi: number;
}

export type EnvironmentSensorData = {
    temperature: number;
    temperatureError: boolean;
    humidity: number;
    batteryMilliVolts: number;
    adcLight: number;
    sleepTimeInSeconds: number;
    timestampDelta: number;
    receivedAt: number;
    rssi: number;
}

export type EnvironmentSensorState = {
    sensorId: number;
    displayName: string;
    sleepTimeInSeconds: number;
    sensorData?: EnvironmentSensorData;
    firmwareUpgradeState?: FirmwareUpgradeState;
    firmwareVersion: number;
    firmwareUpgradeScheduled: boolean;
    resetScheduled: boolean;
    timeAdjustmentSchedule: boolean;
}

export type FirmwareInfo = {
    filename: string;
    size: number;
}


export type EnvironmentSensorResponse = {
    sensors: EnvironmentSensorState[];
    firmwareInfo?: FirmwareInfo;
}

export type EnvironmentSensorEventType = "update"

export type EnvironmentSensorEvent = {
    type: EnvironmentSensorEventType;
    sensors: EnvironmentSensorState[];
    firmwareInfo?: FirmwareInfo;
}

File: src/main/kotlin/com/dehnes/smarthome/api/dtos/EnvironmentSensors.kt
package com.dehnes.smarthome.api.dtos

enum class EnvironmentSensorRequestType {
    getAllEnvironmentSensorData,
    uploadFirmware,

    scheduleTimeAdjustment,
    cancelTimeAdjustment,
    scheduleFirmwareUpgrade,
    cancelFirmwareUpgrade,
    adjustSleepTimeInSeconds,
    scheduleReset,
    cancelReset
}

data class EnvironmentSensorRequest(
    val type: EnvironmentSensorRequestType,
    val sensorId: Int?,
    val firmwareFilename: String?,
    val firmwareBased64Encoded: String?,
    val sleepTimeInSecondsDelta: Int?
)

data class EnvironmentSensorResponse(
    val sensors: List<EnvironmentSensorState>,
    val firmwareInfo: FirmwareInfo?
)

data class FirmwareInfo(
    val filename: String,
    val size: Int
)

data class EnvironmentSensorState(
    val sensorId: Int,
    val displayName: String,
    val sleepTimeInSeconds: Int,
    val sensorData: EnvironmentSensorData?,
    val firmwareUpgradeState: FirmwareUpgradeState?,
    val firmwareVersion: Int,
    val firmwareUpgradeScheduled: Boolean,
    val resetScheduled: Boolean,
    val timeAdjustmentSchedule: Boolean
)

data class FirmwareUpgradeState(
    val firmwareSize: Int,
    val offsetRequested: Int,
    val timestampDelta: Long,
    val receivedAt: Long,
    val rssi: Int
)

data class EnvironmentSensorData(
    val temperature: Long,
    val temperatureError: Boolean,
    val humidity: Long,
    val batteryMilliVolts: Long,
    val adcLight: Long,
    val sleepTimeInSeconds: Int,
    val timestampDelta: Long,
    val receivedAt: Long,
    val rssi: Int
)

enum class EnvironmentSensorEventType {
    update
}

data class EnvironmentSensorEvent(
    val type: EnvironmentSensorEventType,
    val sensors: List<EnvironmentSensorState>,
    val firmwareInfo: FirmwareInfo?
)
